{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Necessary Packages & Enviorments\n",
    "* NCBI Datasets (v18.9.0)\n",
    "    * https://github.com/ncbi/datasets\n",
    "    * conda install -c conda-forge ncbi-datasets-cli\n",
    "* AUGUSTUS (v3.5.0)\n",
    "    * https://github.com/Gaius-Augustus/Augustus\n",
    "    * conda install bioconda::augustus\n",
    "* Prodigal (v2.6.3)\n",
    "    * https://github.com/hyattpd/Prodigal\n",
    "    * conda install bioconda::prodigal\n",
    "* FragGeneScan (v1.32)\n",
    "    * https://github.com/gaberoo/FragGeneScan\n",
    "    * Compiled using GitHub instructions\n",
    "* GFFCompare (v0.12.10)\n",
    "    * https://github.com/gpertea/gffcompare\n",
    "    * conda install bioconda::gffcompare\n",
    "* subprocess \n",
    "    * import subprocess\n",
    "\n",
    "NOTE: The below code assumes that when you run the spcified section, if you created a unique conda enviorment for that package, that enviorment is activated. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Download Data\n",
    "* We will be using the datasets command to download all assemblies and .gff files for the following organisms and associated GenBank accession numbers. The file will be downloaded as .zip files, which will need to be unzipped (which can be accomplished using `unzip \"zip_*\"`). This will cause some messages as unzipping all files will lead to overriding of the README file. Type yes for all, this will then create the directory `ncbi_dataset` which will have a subfolder called `data` where each assembly and GFF file lives for the specified accession numbers. I perfered to move the data folder to the directory that I was doing my work in to make the path less confusing, but this is up to you. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "| Species                   | Accession #        | GC Content | AUGUSTUS Bacteria Reference                                               | Bacteria GC |\n",
    "|----------------------------|--------------------|-------------|----------------------------------------------------------------------|--------------|\n",
    "| Nanoarchaeum equitans      | GCA_000008085.1    | 31.5%       | Staphylococcus aureus (*s_aureus*)                                   | 32.7%        |\n",
    "| Methanococcus maripaludis  | GCA_002945325.1    | 33.0%       | Staphylococcus aureus (*s_aureus*)                                   | 32.7%        |\n",
    "| Sulfolobus solfataricus    | GCA_003852155.1    | 36.0%       | Thermoanaerobacter tengcongensis (*thermoanaerobacter_tengcongensis*)| 37.5%        |\n",
    "| Ferroplasma acidiphilum    | GCA_002078355.1    | 36.5%       | Thermoanaerobacter tengcongensis (*thermoanaerobacter_tengcongensis*)| 37.5%        |\n",
    "| Cuniculiplasma divulgatum  | GCA_900083515.1    | 37.5%       | Thermoanaerobacter tengcongensis (*thermoanaerobacter_tengcongensis*)| 37.5%        |\n",
    "| Pyrococcus furiosus        | GCA_008245085.1    | 41.0%       | Streptococcus pneumoniae (*s_pneumoniae*)                            | 39.5%        |\n",
    "| Ferroglobus placidus       | GCA_000025505.1    | 44%         | Streptococcus pneumoniae (*s_pneumoniae*)                            | 39.5%        |\n",
    "| Thermococcus kodakarensis  | GCA_000009965.1    | 52%         | Escherichia coli K-12 (*E_coli_K12*)                                 | 51%          |\n",
    "| Natronomonas pharaonis     | GCA_000026045.1    | 63%         | Burkholderia pseudomallei (*b_pseudomallei*)                         | 68.0%        |\n",
    "| Halorubrum ezzemoulense    | GCA_004126515.1    | 66.5%       | Burkholderia pseudomallei (*b_pseudomallei*)                         | 68.0%        |\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "\n",
    "download_list = [\"GCA_000008085.1\", \n",
    "                 \"GCA_002945325.1\", \n",
    "                 \"GCA_003852155.1\", \n",
    "                 \"GCA_002078355.1\", \n",
    "                 \"GCA_900083515.1\", \n",
    "                 \"GCA_008245085.1\", \n",
    "                 \"GCA_000025505.1\", \n",
    "                 \"GCA_000009965.1\",\n",
    "                 \"GCA_000026045.1\", \n",
    "                 \"GCA_004126515.1\"]\n",
    "\n",
    "\n",
    "for item in download_list:\n",
    "    file_name = \"zip_\" + item\n",
    "    print(file_name)\n",
    "    command = f'datasets download genome accession {item} --include gff3,genome --filename {file_name}'\n",
    "    print(command)\n",
    "    subprocess.run(command, shell=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Process Data\n",
    "* The GFF files that we downloaded include information that we don't need (i.e they include CDS and gene information, where for the scope of this project we are only focusing on CDS), this means that we need to parse them in order to get only the information we are intrested in. This will be done using the `grep` command. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for item in download_list: #for each organism \n",
    "    data_path = \"data/\" + item + \"/genomic.gff\" #path to where original gff file is\n",
    "    folder_path = \"data/\" + item + \"/\" #path to folder for output\n",
    "    #print(data_path)\n",
    "    command = f'grep \"CDS\" {data_path} > {folder_path}{item}_CDS.gff' #grab all the CDS and put them into a CDS.gff file\n",
    "    #print(command)\n",
    "    subprocess.run(command, shell=True) #running command in terminal\n",
    "    add_line_command = f\"sed -i '1i##gff-version 3' {folder_path}{item}_CDS.gff\" #.gff files require this as the first line, so add it in\n",
    "    #print(add_line_command)\n",
    "    subprocess.run(add_line_command, shell=True)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AUGUSTUS Analysis\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AUGUSTUS was initally make for eukaryotic speices, but since then it has expanded to support some prokaryotic organisms. Unlike the other two programs, AUGUSTUS requires a model. There are two ways that I have ran AUGUSTUS. First I have run all genomes with the only archaea species for which it has a model: Sulfolobus solfataricus. AUGUSTUS also has 5 other prokaryotic (bacteria) species. To see if having a GC content more closely resembling the target species makes a difference in the result, rather than the same Domain, I will choose the model bacterial species with the closest GC content and run a test with that as well (as outlined in the table above). For each run, I have run it with the species, -gff3=on as to get it in the correct output format, and --genemodel=intronless as we are doing prokaryotes. Apart from this, nothing other then the assembly file was inputed. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using Bacteria Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for each run, use the table above that we have listed with the bacteria models, assign it to a list\n",
    "species_model_list = [\"s_aureus\",\n",
    "                      \"s_aureus\",\n",
    "                      \"thermoanaerobacter_tengcongensis\",\n",
    "                      \"thermoanaerobacter_tengcongensis\",\n",
    "                      \"thermoanaerobacter_tengcongensis\",\n",
    "                      \"s_pneumoniae\",\n",
    "                      \"s_pneumoniae\",\n",
    "                      \"E_coli_K12\",\n",
    "                      \"b_pseudomallei\",\n",
    "                      \"b_pseudomallei\"]\n",
    "\n",
    "os.mkdir(\"augustus_bacteria_output\") #make an output folder for the augustus bacteria results\n",
    "\n",
    "for item,species in zip(download_list,species_model_list): #for each organism, and it's associated bacteria species model\n",
    "    print(f'Running AUGUSTUS for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'augustus_bacteria_output/{item}') #make a folder for it in the augustus bacteria results\n",
    "    \n",
    "    path_to_fna = \"data/\" + item + \"/*.fna\" #path to where the .fna file exits\n",
    "    folder_path = \"augustus_bacteria_output/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    output_file = f'augustus_{item}.gff' #name of output file with raw results\n",
    "    CDS_output_file = f'augustus_{item}_CDS.gff' #name of output file with CDS only results\n",
    "    \n",
    "    augustus_command = f'/usr/bin/time --verbose augustus --species={species} --gff3=on --genemodel=intronless {path_to_fna} > {folder_path}{output_file}'\n",
    "    subprocess.run(augustus_command, shell=True) #command to do augustus using specified species, gff3 format, and intronless model as we are doing prokaryotes.\n",
    "    #print(augustus_command)\n",
    "    \n",
    "    grep_command = f'grep \"AUGUSTUS\\tCDS\" {folder_path}{output_file} > {folder_path}{CDS_output_file}'\n",
    "    subprocess.run(grep_command, shell=True) #command to get just CDS results\n",
    "    #print(grep_command)\n",
    "   \n",
    "    sed_comman = f\"sed -i '1i##gff-version 3' {folder_path}{CDS_output_file}\"\n",
    "    subprocess.run(sed_comman, shell=True) #format it in gff3\n",
    "    #print(sed_comman)\n",
    "    print(\"\\n\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Using Archaea Model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#since there is only one Archaea species, we don't need a list like we did for bacteria\n",
    "\n",
    "os.mkdir(\"augustus_archaea_output\") #make an output folder for the augustus bacteria results\n",
    "\n",
    "for item in download_list: #for each organism\n",
    "    print(f'Running AUGUSTUS for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'augustus_archaea_output/{item}') #make a folder for it in the augustus bacteria results\n",
    "    \n",
    "    path_to_fna = \"data/\" + item + \"/*.fna\" #path to where the .fna file exits\n",
    "    folder_path = \"augustus_archaea_output/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    output_file = f'augustus_{item}.gff' #name of output file with raw results\n",
    "    CDS_output_file = f'augustus_{item}_CDS.gff' #name of output file with CDS only results\n",
    "    \n",
    "    augustus_command = f'/usr/bin/time --verbose augustus --species=sulfolobus_solfataricus --gff3=on --genemodel=intronless {path_to_fna} > {folder_path}{output_file}'\n",
    "    subprocess.run(augustus_command, shell=True) #command to do augustus using specified species, gff3 format, and intronless model as we are doing prokaryotes.\n",
    "    #print(augustus_command)\n",
    "    \n",
    "    grep_command = f'grep \"AUGUSTUS\\tCDS\" {folder_path}{output_file} > {folder_path}{CDS_output_file}'\n",
    "    subprocess.run(grep_command, shell=True) #command to get just CDS results\n",
    "    #print(grep_command)\n",
    "   \n",
    "    sed_comman = f\"sed -i '1i##gff-version 3' {folder_path}{CDS_output_file}\"\n",
    "    subprocess.run(sed_comman, shell=True) #format it in gff3\n",
    "    #print(sed_comman)\n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prodigal Analysis\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Prodigal requires no model. In running it I just used the default command, specifying the input/output paths as well as the perferred format (gff)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.mkdir(\"prodigal_output\") #make an output folder for the augustus bacteria results\n",
    "\n",
    "for item in download_list: #for each organism\n",
    "    print(f'Running Prodigal for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'prodigal_output/{item}') #make a folder for it in the prodigal results\n",
    "    \n",
    "    path_to_fna = \"data/\" + item + \"/*.fna\" #path to where the .fna file exits\n",
    "    folder_path = \"prodigal_output/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    output_file = f'prodigal_{item}_CDS.gff' #name of output file with results\n",
    "    \n",
    "    print(output_file)\n",
    "    \n",
    "    prodigal_command = f'/usr/bin/time --verbose prodigal -i {path_to_fna} -o {folder_path}{output_file} -f gff'\n",
    "    #print(prodigal_command)\n",
    "    subprocess.run(prodigal_command, shell=True) #command to do prodigal\n",
    "    \n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FragGeneScan Analysis\n",
    "\n",
    "* This will have to be done in the directory that you made your FragGeneScan (i.e where the run_FragGeneScan.pl file exist). FragGeneScan also requires no model. In addition to the input and output paths, I also specified that the assembly genomes we are using are complete. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "os.mkdir(\"fraggenescan_output\") #make an output folder for the augustus bacteria results\n",
    "\n",
    "for item in download_list: #for each organism\n",
    "    print(f'Running FragGeneScan for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'fraggenescan_output/{item}') #make a folder for it in the prodigal results\n",
    "    \n",
    "    path_to_fna = \"../data/\" + item + \"/*.fna\" #path to where the .fna file exits, not full\n",
    "    \n",
    "    echo_path = subprocess.run(f\"echo {path_to_fna}\", shell=True, capture_output=True, text=True) #get full path\n",
    "    full_path_to_fna = echo_path.stdout.strip() #strip and format\n",
    "    \n",
    "    folder_path = \"fraggenescan_output/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    output_file = f'fraggenescan_{item}_CDS' #name of output file with results\n",
    "    \n",
    "    print(output_file)\n",
    "    \n",
    "    fraggenescan_command = f'/usr/bin/time --verbose ./run_FragGeneScan.pl -genome={full_path_to_fna} -out={folder_path}{output_file} -complete=1 -train=complete'\n",
    "    \n",
    "    print(fraggenescan_command)\n",
    "    subprocess.run(fraggenescan_command, shell=True) #command to do fraggenescan\n",
    "    \n",
    "    print(\"\\n\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Compare Output with RefSeq GFF File Using GFFCompare"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "######## AUGUSTUS COMPARISONS - Bacteria ########\n",
    "\n",
    "os.mkdir(\"augustus_bacteria_compare\") #make an output folder for the augustus bacteria comparison\n",
    "\n",
    "for item in download_list: #for each organism, and it's associated bacteria species model\n",
    "    print(f'Running AUGUSTUS bacteria compare for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'augustus_bacteria_compare/{item}') #make a folder for it in the augustus bacteria results\n",
    "    \n",
    "    path_to_true_gff = \"data/\" + item + \"/*_CDS.gff\" #path to where the .fna file exits\n",
    "    path_to_test_gff = \"augustus_bacteria_output/\" + item + \"/*_CDS.gff\"\n",
    "    \n",
    "    folder_path = \"augustus_bacteria_compare/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    \n",
    "    gff_command = f'/usr/bin/time --verbose gffcompare -r {path_to_true_gff} {path_to_test_gff}'\n",
    "    subprocess.run(gff_command, shell=True) #command to do comparisons\n",
    "    print(gff_command)\n",
    "    \n",
    "    move_ouput_command = f'mv gffcmp.* {folder_path}'\n",
    "    subprocess.run(move_ouput_command, shell=True) #command to do move output\n",
    "    print(move_ouput_command)\n",
    "\n",
    "    print(\"\\n\")\n",
    "    \n",
    "\n",
    "######## AUGUSTUS COMPARISONS - Archaea ########\n",
    "\n",
    "os.mkdir(\"augustus_archaea_compare\") #make an output folder for the augustus archaea comparison\n",
    "\n",
    "for item in download_list: #for each organism, and it's associated archaea species model\n",
    "    print(f'Running AUGUSTUS archaea compare for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'augustus_archaea_compare/{item}') #make a folder for it in the augustus archaea results\n",
    "    \n",
    "    path_to_true_gff = \"data/\" + item + \"/*_CDS.gff\" #path to where the .fna file exits\n",
    "    path_to_test_gff = \"augustus_archaea_output/\" + item + \"/*_CDS.gff\"\n",
    "    \n",
    "    folder_path = \"augustus_archaea_compare/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    \n",
    "    gff_command = f'/usr/bin/time --verbose gffcompare -r {path_to_true_gff} {path_to_test_gff}'\n",
    "    subprocess.run(gff_command, shell=True) #command to do comparisons\n",
    "    print(gff_command)\n",
    "    \n",
    "    move_ouput_command = f'mv gffcmp.* {folder_path}'\n",
    "    subprocess.run(move_ouput_command, shell=True) #command to do move output\n",
    "    print(move_ouput_command)\n",
    "\n",
    "    print(\"\\n\")\n",
    "\n",
    "\n",
    "######## PRODIGAL COMPARISONS ########\n",
    "\n",
    "os.mkdir(\"prodigal_compare\") #make an output folder for the Prodigal comparison\n",
    "\n",
    "for item in download_list: #for each organism\n",
    "    print(f'Running Prodigal compare for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'prodigal_compare/{item}') #make a folder for it in the Prodigal results\n",
    "    \n",
    "    path_to_true_gff = \"data/\" + item + \"/*_CDS.gff\" #path to where the .fna file exits\n",
    "    path_to_test_gff = \"prodigal_output/\" + item + \"/*_CDS.gff\"\n",
    "    \n",
    "    folder_path = \"prodigal_compare/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    \n",
    "    gff_command = f'/usr/bin/time --verbose gffcompare -r {path_to_true_gff} {path_to_test_gff}'\n",
    "    subprocess.run(gff_command, shell=True) #command to do comparisons\n",
    "    print(gff_command)\n",
    "    \n",
    "    move_ouput_command = f'mv gffcmp.* {folder_path}'\n",
    "    subprocess.run(move_ouput_command, shell=True) #command to do move output\n",
    "    print(move_ouput_command)\n",
    "\n",
    "    print(\"\\n\")\n",
    "    \n",
    "\n",
    "######## FRAGGENESCAN COMPARISONS ########\n",
    "\n",
    "os.mkdir(\"fraggenescan_compare\") #make an output folder for the fraggenescan comparison\n",
    "\n",
    "for item in download_list: #for each organism\n",
    "    print(f'Running FragGeneScan compare for {item}\\n') #print what organism we are doing analysis for, this will be useful when we save the terminal output to compare time/computational intensity later.\n",
    "    os.mkdir(f'fraggenescan_compare/{item}') #make a folder for it in the fraggenescan results\n",
    "    \n",
    "    path_to_true_gff = \"data/\" + item + \"/*_CDS.gff\" #path to where the .fna file exits\n",
    "    path_to_test_gff = \"fraggenescan_output/\" + item + \"/*_CDS.gff\"\n",
    "    \n",
    "    folder_path = \"fraggenescan_compare/\" + item + \"/\" #path to the folder that the output will be in\n",
    "    #print(path_to_fna)\n",
    "    \n",
    "    gff_command = f'/usr/bin/time --verbose gffcompare -r {path_to_true_gff} {path_to_test_gff}'\n",
    "    subprocess.run(gff_command, shell=True) #command to do comparisons\n",
    "    print(gff_command)\n",
    "    \n",
    "    move_ouput_command = f'mv gffcmp.* {folder_path}'\n",
    "    subprocess.run(move_ouput_command, shell=True) #command to do move output\n",
    "    print(move_ouput_command)\n",
    "\n",
    "    print(\"\\n\")\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Manual Curation of Comparisons Spreadsheet\n",
    "After all comparisons were done, the output was manually examined to create a spreadsheet that included the base level sensitivity and precision, transcript level sensitivity and precision, locus level sensitivity and precision, number of matching transcripts, matching loci, missed loci, and novel loci for each run. From the time command output, how long each run took (in seconds), as well as the maximum resident set size (KB), was added to the spreadsheet. This information was supplemented by also adding the GC content and base length of each genome. "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
